pkg/store/postings_codec.go

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package store
+
+import (
+	"bytes"
+
+	"github.com/golang/snappy"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/encoding"
+	"github.com/prometheus/prometheus/tsdb/index"
+)
+
+const (
+	// These headers should not be a prefix of each other.
+	codecHeaderRaw    = "diff+varint+raw"
+	codecHeaderSnappy = "diff+varint+snappy"
+)
+
+// isDiffVarintEncodedPostings returns true, if input looks like it has been encoded by diff+varint(+snappy) codec.
+func isDiffVarintEncodedPostings(input []byte) bool {
+	return bytes.HasPrefix(input, []byte(codecHeaderRaw)) || bytes.HasPrefix(input, []byte(codecHeaderSnappy))
+}
+
+// diffVarintEncode encodes postings into diff+varint representation and optional snappy compression.
+func diffVarintEncode(p index.Postings, useSnappy bool) ([]byte, error) {
+	buf := encoding.Encbuf{}
+
+	// If we're returning raw data, write the header to the buffer, and then return buffer directly.
+	if !useSnappy {
+		buf.PutString(codecHeaderRaw)
+	}
+
+	prev := uint64(0)
+	for p.Next() {
+		v := p.At()
+		if v < prev {
+			// Postings entries must be in increasing order.
+			return nil, errors.Errorf("decreasing entry, val: %d, prev: %d", v, prev)
+		}
+
+		buf.PutUvarint64(v - prev)
+		prev = v
+	}
+
+	if p.Err() != nil {
+		return nil, p.Err()
+	}
+
+	if !useSnappy {
+		// When not using Snappy, buffer already has the correct header.
+		return buf.B, nil
+	}
+
+	// Make result buffer large enough to hold our header and compressed block.
+	resultBuf := make([]byte, len(codecHeaderSnappy)+snappy.MaxEncodedLen(buf.Len()))
+	copy(resultBuf, codecHeaderSnappy)
+
+	compressed := snappy.Encode(resultBuf[len(codecHeaderSnappy):], buf.B)
+
+	// Slice result buffer based on compressed size.
+	resultBuf = resultBuf[:len(codecHeaderSnappy)+len(compressed)]
+	return resultBuf, nil
+}
+
+func diffVarintDecode(input []byte) (index.Postings, error) {
+	compressed := false
+	headerLen := 0
+	switch {
+	case bytes.HasPrefix(input, []byte(codecHeaderRaw)):
+		headerLen = len(codecHeaderRaw)
+	case bytes.HasPrefix(input, []byte(codecHeaderSnappy)):
+		headerLen = len(codecHeaderSnappy)
+		compressed = true
+	default:
+		return nil, errors.New("header not found")
+	}
+
+	raw := input[headerLen:]
+	if compressed {
```

Comment:
just put it in `case bytes.HasPrefix(input, []byte(codecHeaderSnappy)):`? No need for `compressed` bool

Reply:
By using bool flag, we can avoid `HasPrefix` comparison which is costly. But I think I'll remove non-snappy version completely, as you suggested in other comment.

Reply:
yes please (: